link-grammar/parse/preparation.c

```diff
@@ -195,20 +195,40 @@ static void create_wildcard_word_disjunct_list(Sentence sent,
  */
 void prepare_to_parse(Sentence sent, Parse_Options opts)
 {
-	size_t i;
-
 	if (IS_GENERATION(sent->dict))
 		create_wildcard_word_disjunct_list(sent, opts);
 
 	build_sentence_disjuncts(sent, opts->disjunct_cost, opts);
+
+
+#ifdef NEED_THESE_FOR_NULL_AND_PANIC_PARSING
```

Comment:
We can add a parse option to indicate that a second pass is not expected. In that case, we can not only free the `Exp_pool` but immediately delete all the pools that are now pool-reused.

If a second pass happens anyway (disregarding the value of the said parse option) and there are no expressions, we can just return an error.

Reply:
There can only be a second pass if either
```
 null         Allow null links                                  1 (On)
 panic        Use of "panic mode"                               1 (On)
```
If both are off, then there's no second pass, right?

Reply:
Yes. But this is a link-parser convention only. The library doesn't know about it. For example, if the library gets a `sentence_parse()`  request with both `min_null_count==0` and `max_null_count==0`, it doesn't know whether a subsequent call with, e.g., `min_null_count==1` and `max_null_count=sent_length` will follow. So maybe we can add a `free_memory` or even a `batch_mode`  parse option.

Reply:
Ah, yes. I forgot what this was about. Perhaps a new `struct Parse_Optiions_1 of `bool single_pass` ?